Add comprehensive documentation for Ovum bytecode.

bialger opened the pull request:
Describe ovum bytecode command descriptions, syntax, control flow structures, class declarations, and examples of usage. This enhances the language reference and provides clear guidance for developers.

bialger committed:
docs: add comprehensive documentation for Ovum bytecode.

This includes command descriptions, syntax, control flow structures, class declarations, and examples of usage. This enhances the language reference and provides clear guidance for developers.

sashbek commented on reference/bytecode.md:
````diff
@@ -0,0 +1,239 @@
+```oil
+{
+    // Condition block - pushes Bool value to stack
+    {
````
по логике получается структура if-then-else. Предлагаю переделать так:
```
if {
    condition
} then {
    ex code 1
} else if {
    condition
} then {
    ex code 2
} else {
    ex code 3
}
```

sashbek commented on reference/bytecode.md:
```diff
@@ -0,0 +1,239 @@
+    SetLocal 0
+    
+    // While loop
+    while {
```
Аналогично кажется лучше будет
```
while {
    condition
} then {
    ex loop iteration code
}
```

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+* `FloatSubtract` - Subtracts two floats (a - b)
+* `FloatMultiply` - Multiplies two floats (a * b)
+* `FloatDivide` - Divides two floats (a / b)
+* `FloatModulo` - Computes remainder (a % b)
```
Необходимо пояснение, зачем такая команда? Реализация не оправдана, как мне кажется.

bialger replied:
Команда не нужна, уберу

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+* `CopyFile` - Copies file
+* `ListDirectory` - Lists directory contents
+* `GetCurrentDirectory` - Gets current working directory
+* `ChangeDirectory` - Changes current directory
```
добавить "закрыть файл"

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+
+## Process Control
+
+* `Sleep` - Sleeps for specified milliseconds
```
Для единообразия предлагаю переделать в SleepMs

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+
+* `Sleep` - Sleeps for specified milliseconds
+* `SleepNs` - Sleeps for specified nanoseconds
+* `Exit` - Terminates process with exit code
```
с каким exit code? откуда он берется?

bialger replied:
Как и с остальными инструкциями: аргументы берутся со стека, результат если есть кладётся туда. Стоит ли везде это прописывать?

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+* `Exit` - Terminates process with exit code
+* `GetProcessId` - Gets current process ID
+* `GetEnvironmentVariable` - Gets environment variable
+* `SetEnvironmentVariable` - Sets environment variable
```
откуда берется имя и значение EnvVariable?

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+* `GetMemoryUsage` - Gets current memory usage
+* `GetPeakMemoryUsage` - Gets peak memory usage
+* `ForceGarbageCollection` - Forces garbage collection
+* `GetProcessorCount` - Gets number of CPU cores
```
useless же, зачем нам ядра процессора, у нас нет многопоточки

bialger replied:
На будущее

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+
+## Memory Management
+
+* `Allocate <size>` - Allocates memory block
```
... и помещает его указатель на стек? или указатель и размер, чтоб не забыли? Нужно описать подробнее.

bialger replied:
Есть у меня честно скажу ощущение, что эти команды нам делать __запрещено__ - это ручное управление памятью. Я их удалю

sashbek commented on reference/bytecode_examples.md:
мне кажется, лучше прописать примеры по типу 
`ovum code -> oil bytecode`
т. е. добавить к каждому примеру программу на ovum которая транслируется в такой байткод

sashbek commented on reference/bytecode_commands.md:
```diff
@@ -0,0 +1,206 @@
+* `BoolNot` - Logical NOT (!a)
+* `BoolXor` - Logical XOR (a ^ b)
+
+## Bitwise Operations
```
`Byte` мы обрабатываем через Int? Возможно, стоит игнорировать знак в этих операциях? Если мы игнорируем знак, то нужно это упомянуть.

bialger replied:
Да, byte в отличие от char беззнаковый. Распишу

sashbek reviewed (commented):
В целом все +- ок, нужно подробнее расписать `memory management`, `Error Handling`и прописать более явно все команды. 
Пример:
Не оч понятно: 
`GetOsName` - Gets operating system name
Лучше:
`GetOsName` - Gets operating system name and push it on stack
или:
`GetOsName` - Gets operating system name and adds it to local vars with id 777

bialger committed:
docs: enhance Ovum bytecode documentation with new commands and examples

Added descriptions for byte operations, comparison operations, and bitwise operations. Updated example programs to illustrate the compilation of Ovum source code to OIL bytecode, including class declarations and pure functions. Improved clarity on control flow structures in bytecode.

bialger committed:
chore: add .gitignore file to exclude IDE and editor configuration files

This commit introduces a .gitignore file to prevent version control of IDE-specific files, including those from VSCode and IntelliJ IDEA.

bialger committed:
refactor: update bytecode examples for consistency and clarity

This commit modifies the bytecode examples to use consistent variable declarations with 'val' instead of 'let', updates method visibility to 'public', and ensures proper usage of the 'sys' namespace for printing and mathematical functions. These changes enhance readability and maintainability of the examples.

bialger committed:
docs: update bytecode documentation with new commands and examples

This commit adds the `FloatSqrt` command to the bytecode commands documentation and introduces the `Unwrap` command for converting values between primitive and fundamental types. Additionally, it refines the bytecode examples for clarity, ensuring consistent variable declarations and method visibility. These enhancements improve the overall readability and usability of the documentation.

bialger commented:
@codex review

sashbek reviewed (approved):
норм.
